### PR TITLE
[target-allocator] Introduce "per node" allocation strategy to target allocator

### LINKED ISSUE #1828: TargetAllocator for daemonset
Ideally, when deployed as a daemonset, collectors would scrape the pods running on the node
local to the daemonset instance. It's not obvious from the docs or configs as to if that is achievable
with the current targetAllocator strategies.

Am I missing something? Would you accept a PR that implements this?

### PR DESCRIPTION
**Pull request by matej-g**
**Description:**
Resolves #1828.

Introduces new allocation strategy "per node" that is intended to distribute targets to collectors on the basis of the node on which the targets reside, i.e. each collector will be instructed to scrape targers running on the collector's node.

This strategy is intended to be only used with a collector in `deamonset` mode (running as agent on each node). It's not suitable for other modes.

**Link to tracking Issue:** #1828 

**Testing:**
- Added unit tests
- Also did manual testing

**Documentation:** Updated API documentation.

### CODE REVIEW

File: cmd/otel-allocator/allocation/per_node.go
Comment (matej-g) [inline cmd/otel-allocator/allocation/per_node.go:32]: One doubt I'm having is how to make the label matching more robust, I'm mainly thinking about:
- Which service discoveries should we be targeting here? For service / pod monitors, we should be able depend on `__meta_kubernetes_pod_node_name`, but I wonder if there might be cases where this label is not available (perhaps due to using different role for Kubernetes SD or using different SD? Has anyone encountered people using other SDs?)
- How to handle targets that we might not match against any node (such as static configs), currently I'm skipping those. I'm thinking if that's fine (maybe should be documented?) or we want to have some type of fallback to distribute such targets, if that makes sense?

Comment (swiatekm-sumo) [reply cmd/otel-allocator/allocation/per_node.go:33]: For the labels, we are constrained by the fact that the value needs to match the collector's Node field. So we can't do much better than Kubernetes Service Discovery, or at least not without significant additional complexity. So what I'd do there is just check all the Node fields from SD:

* `__meta_kubernetes_node_name`
* `__meta_kubernetes_pod_node_name`
* `__meta_kubernetes_endpoint_node_name`

The problem of assigning targets without a Node doesn't have an obvious solution. I can see the following options:

* Ignore these targets
* Assign them randomly
* Assign them according to some deterministic scheme (could be what the consistent hashing strategy uses)

We should discuss this in the issue, rather than here imo.

**Comment by swiatekm-sumo** [comment]
> I also noticed we are reusing a lot of code between allocation strategies. I was thinking about adjusting the abstraction (perhaps allocator does not need to be an interface, we could maybe have more specific interface that would implement only methods that differ across strategies). But I did not want to introduce too many changes at once.

Yeah, we can do that afterwards in a separate PR.

File: cmd/otel-allocator/allocation/per_node.go
Comment (jaronoff97) [inline cmd/otel-allocator/allocation/per_node.go:42]: thought: i think it may be more efficient if we were to store collectors as a map from nodename -> collector rather than the collector name. I think this would make a few operations like removing a collector a bit less efficient, but it would make target assignment much more efficient. This is an acceptable tradeoff as i would imagine targets change more often than collectors

**Review by jaronoff97 — APPROVED** [review]
Overall I really like the improvement here. I have one suggestion which may help with making it more efficient, interested in hearing your thoughts. Thank you!

**Comment by swiatekm-sumo** [comment]
> Thanks for the pointer @swiatekm-sumo, I have added an E2E test and also reorganized the tests a bit so that TA E2E test can be run separately under separate target (hopefully you don't mind 🙂).

Is there a reason to do this? If you only want the target allocator tests, you can use kuttl's `--test` flag to select them.

And if you really want to, you should make sure they're run in the CI as well.

File: cmd/otel-allocator/allocation/per_node.go
Comment (matej-g) [reply cmd/otel-allocator/allocation/per_node.go:41]: Makes sense to me and it's a simple optimization, made the change

**Comment by matej-g** [comment]
> Is there a reason to do this? If you only want the target allocator tests, you can use kuttl's `--test` flag to select them.
> 
> And if you really want to, you should make sure they're run in the CI as well.

I think it makes sense with respect to how other test cases are organized, since we have usually different test directories for different components (OpAMP bridge, auto-instrumentation). But you're right, it was missing from the list in the action config, I fixed that now.

File: tests/e2e-targetallocator/targetallocator-kubernetessd/00-install.yaml
Comment (jaronoff97) [inline tests/e2e-targetallocator/targetallocator-kubernetessd/00-install.yaml:102]: I may be missing something but don't we want to test the new strategy here? Wouldn't our final teststep in 01-install.yaml fail without installing the TA?

Comment (jaronoff97) [reply tests/e2e-targetallocator/targetallocator-kubernetessd/00-install.yaml:102]: wouldn't we expect [this](https://github.com/open-telemetry/opentelemetry-operator/pull/2430/files#diff-969909de4d71fcaae234fd2a1ce3bd7c03b726126e2443bb7dc04c8a5145b97eR49) to fail then? And also for [this target allocator](https://github.com/open-telemetry/opentelemetry-operator/pull/2430/files#diff-fbf022ca14c8b039493f038feac0716db643ac051a41ccb76bac47cb3bcefeb8R35-R38) section to be empty?

Comment (jaronoff97) [reply tests/e2e-targetallocator/targetallocator-kubernetessd/00-install.yaml:102]: should we require that daemonset mode has to run with the new strategy? right now we require the new strategy has daemonset but not the reverse. I don't think anything could go horribly wrong if you run with least-weighted and daemonset, but i would worry that too much node cycling would cause poorly balanced targets.

Comment (matej-g) [reply tests/e2e-targetallocator/targetallocator-kubernetessd/00-install.yaml:102]: I added a validation for this as well

**Comment by matej-g** [comment]
Hey folks,
This should hopefully be ready for final reviews.

I was testing this a bit more and got a report about issue when collector pods are changed (e.g. added / removed nodes or daemonset being re-created), whereas after such change TA was not allocating target anymore. After a deep dive I finally figured this is because the node name in the pod spec is not always available when we get the pod data from watcher. To solve this, I added a condition to restart the watcher, until we receiver node name on collector pods. Let me know if this makes sense to you. Thanks.

File: cmd/otel-allocator/allocation/per_node.go
Comment (jaronoff97) [inline cmd/otel-allocator/allocation/per_node.go:193]: I think we should just make this a gauge metric that counts the length of the unassigned map or even just count the unassigned as an int and record that.

**Review by jaronoff97 — COMMENTED** [review]
one comment, one suggestion. Let me know your thoughts on the suggestion

File: apis/v1alpha1/collector_webhook.go
Comment (matej-g) [reply apis/v1alpha1/collector_webhook.go:240]: I think that's a good idea idea at this point, I'll put TA validations into a separate method.

File: cmd/otel-allocator/allocation/per_node.go
Comment (matej-g) [reply cmd/otel-allocator/allocation/per_node.go:193]: You mean in addition to logs or no logs, just metrics? I don't have strong opinion honestly, either solution works me (as long as it's not too noisy).

Comment (jaronoff97) [reply cmd/otel-allocator/allocation/per_node.go:193]: I think maybe a debug log and metric of just the count of unassigned jobs. Otherwise it may get expensive to iterate through the entire list

**Comment by pavolloffay** [comment]
@matej-g what is the status of this PR? It needs to be rebased.

**Comment by matej-g** [comment]
> @matej-g what is the status of this PR? It needs to be rebased.

Hey, planning to address remaining feedback by Monday, after that hopefully @jaronoff97 can give final seal of approval

File: apis/v1alpha1/collector_webhook.go
Comment (jaronoff97) [inline apis/v1alpha1/collector_webhook.go:235]: we should append the warnings from the above call before returning

File: cmd/otel-allocator/allocation/per_node.go
Comment (jaronoff97) [inline cmd/otel-allocator/allocation/per_node.go:192]: this can just become an integer that's incremented on line 186, i don't think there's a need for a map, unless we care about deduping (which should happen already in the filter strategy)

File: apis/v1alpha1/collector_webhook.go
Comment (matej-g) [reply apis/v1alpha1/collector_webhook.go]: D'oh... thanks, fixing it

File: cmd/otel-allocator/allocation/per_node.go
Comment (matej-g) [reply cmd/otel-allocator/allocation/per_node.go]: I was wondering if there can be a scenario where we have multiple targets per job, that cannot be assigned. But I see the point, I simplified this to just count the number of targets, that should hopefully give enough info when debugging.

File: cmd/otel-allocator/allocation/per_node_test.go
Comment (swiatekm-sumo) [inline cmd/otel-allocator/allocation/per_node_test.go:30]: I'd like to see a few more tests for edge cases. No collectors, no targets, those kinds of conditions. You can probably copy those from consistent-hashing. Ideally we'd have a strategy-agnostic set of tests, but I'm not going to ask that you add those to this (already large and complex) PR.

**Review by swiatekm-sumo — COMMENTED** [review]
Looks good overall. I've left some comments around trying to simplify the bookeeping code in the strategy implementation, but I don't think any of it is necessarily blocking - if you'd like to clean it up in a following PR, I'm ok with that as well.

**Comment by matej-g** [comment]
Hey @swiatekm-sumo, thanks for the detailed review. I'm planning to address all of it, not sure how fast I will get to it. I'm happy to do it in this PR or the next one, whatever will be easier for maintainers to review. 

I'm also in no rush to merge this (I'm using a fork for my needs), but I'm not sure if there are other users waiting for this feature.

**Comment by matej-g** [comment]
Hey folks,
Finally managed to get back to this, I tried to address all the feedback:
- I simplified the code as suggested where it was unnecessarily complicated from the methods taken over from other strategies
- Corrected the documentation where it referenced incorrect information
- Made the unassigned metric a gauge instead of counter
- Added unit test for edge cases (analogous from consistent hashing)

Hopefully this is ready for the final review, thank you @swiatekm-sumo @jaronoff97 🙇

File: cmd/otel-allocator/allocation/per_node.go
Comment (matej-g) [reply cmd/otel-allocator/allocation/per_node.go]: Fixed as suggested

File: cmd/otel-allocator/allocation/strategy.go
Comment (matej-g) [reply cmd/otel-allocator/allocation/strategy.go]: It was intended for both.

I changed to gauge, I'd agree it is more useful. Seems like we use interchangeably counters and gauges for different target number metrics, might be good candidate for some harmonization.

File: cmd/otel-allocator/allocation/per_node_test.go
Comment (matej-g) [reply cmd/otel-allocator/allocation/per_node_test.go:30]: Added tests as suggested.

**Comment by jaronoff97** [comment]
@matej-g thank you! what was the reason for changing the metric type? that may break anyone who has existing dashboards for the TA, we should just rename it too if we do want it to be a gauge

**Review by jaronoff97 — COMMENTED** [review]
Just one concern around the metric type change, otherwise this LGTM.

**Comment by matej-g** [comment]
Hey @jaronoff97, this was based off of the suggestion in https://github.com/open-telemetry/opentelemetry-operator/pull/2430#discussion_r1457269919 - this is a new metric we're introducing with this change, so it should not be breaking

**Review by jaronoff97 — APPROVED** [review]
Ah that's right. Looks good to me! I'll wait on @swiatekm-sumo 's review to merge. Thank you so much!

**Review by swiatekm-sumo — APPROVED** [review]

**Comment by jaronoff97** [comment]
@matej-g thanks for this awesome work!! I really appreciate it.

**Comment by swiatekm-sumo** [comment]
:clap: indeed.

**Comment by matej-g** [comment]
My pleasure, thanks you both for helping me move it over the finish line!

**Pull request merged**
